pydantic/main.py

```diff
@@ -111,7 +112,13 @@ def __setattr__(self, name, value):
             raise ValueError(f'"{self.__class__.__name__}" object has no field "{name}"')
         elif not self.config.allow_mutation:
             raise TypeError(f'"{self.__class__.__name__}" is immutable and does not support item assignment')
-        self.__values__[name] = value
+        if self.config.validate_assignment:
+            value_, error_ = self.fields[name].validate(value, {})
```

Comment:
the second argument to `validate()` should be the other values on the modal in case validation of this field needs to see these values. Here I guess you should use `self.values(exclude={name})`.

You should also add a test for this case.